tracing: Add source IP address to XRay trace

nealp9084 opened the pull request:
Signed-off-by: Neal Patel <nealp9084@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: This PR adds additional fields to the AWS XRay trace related to the client IP address:

* `http.request.client_ip`
* `http.request.x_forwarded_for`

For details, see: https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-http

Risk Level: Low
Testing: Updated unit tests, built on mac os, and built on ubuntu 18.04
Docs Changes: n/a
Release Notes: n/a

lavignes commented:
just a note. Your commit message/title includes "(#1)" since you PR'd this in your own fork. It is a little confusing and I'm surprised it didn't break github's issue parsing since "#N" are treated as links to issues in a repo. In the future, try and remove that that issue number doesn't apply in envoypoxy/envoy.

mattklein123 commented on source/common/tracing/http_tracer_impl.cc:
```diff
@@ -173,6 +173,12 @@ void HttpTracerUtility::finalizeDownstreamSpan(Span& span, const Http::HeaderMap
     span.setTag(Tracing::Tags::get().HttpProtocol,
                 AccessLog::AccessLogFormatUtils::protocolToString(stream_info.protocol()));
 
+    const auto& remoteAddress = stream_info.downstreamDirectRemoteAddress();
```
nit: snake_case for all variable names. Here and elsewhere.

mattklein123 commented on source/common/tracing/http_tracer_impl.cc:
```diff
@@ -173,6 +173,12 @@ void HttpTracerUtility::finalizeDownstreamSpan(Span& span, const Http::HeaderMap
+    const auto& remoteAddress = stream_info.downstreamDirectRemoteAddress();
+    ASSERT(remoteAddress != nullptr);
+    const auto remoteIP = remoteAddress->ip();
+    ASSERT(remoteIP != nullptr);
```
This not guaranteed to be true for UDS downstreams.

nealp9084 commented:
Thanks for the review. I added a test case for UDS, fixed the inconsistent code style, fixed the commit message (and PR title), and rebased against envoyproxy/envoy.

mattklein123 commented on source/common/tracing/http_tracer_impl.cc:
```diff
@@ -173,6 +174,17 @@ void HttpTracerUtility::finalizeDownstreamSpan(Span& span, const Http::HeaderMap
                 AccessLog::AccessLogFormatUtils::protocolToString(stream_info.protocol()));
 
+    const auto& remote_address = stream_info.downstreamDirectRemoteAddress();
+    ASSERT(remote_address != nullptr);
```
nit: del assert which will crash in an obvious way on the next line, same at 182 below.

nealp9084 commented:
Thanks, I removed the assertions in a separate commit.
